cfme/storage/volume.py

```diff
@@ -121,22 +167,88 @@ def is_displayed(self):
 
 
+class VolumeCollection(Navigatable):
+    """Collection object for the :py:class:'cfme.storage.volume.Volume'. """
+
+    def instantiate(self, name, provider):
+        return Volume(name, provider, collection=self)
+
+    def delete(self, *volumes):
+        """Delete one or more Volumes from list of Volumes
+
+        Args:
+            One or Multiple 'cfme.storage.volume.Volume' objects
+        """
+
+        view = navigate_to(self, 'All')
+
+        if view.entities.get_all():
+            for volume in volumes:
+                try:
+                    view.entities.get_entity(volume.name).check()
+                except ItemNotFound:
+                    raise VolumeNotFound("Volume {} not found".format(volume.name))
+
+            view.toolbar.configuration.item_select('Delete selected Cloud Volumes',
+                                                   handle_alert=True)
+
+            for volume in volumes:
+                volume.wait_for_disappear()
+        else:
+            raise VolumeNotFound('No Cloud Volume for Deletion')
+
+
 class Volume(Navigatable):
     # Navigation menu option
     nav = VersionPick({
         Version.lowest(): ['Storage', 'Volumes'],
         '5.8': ['Storage', 'Block Storage', 'Volumes']})
 
-    def __init__(self, name, provider, appliance=None):
-        Navigatable.__init__(self, appliance=appliance)
+    def __init__(self, name, provider, collection=None, appliance=None):
         self.name = name
         # TODO add storage provider parameter, needed for accurate details nav
         # the storage providers have different names then cloud providers
         # https://bugzilla.redhat.com/show_bug.cgi?id=1455270
         self.provider = provider
+        self.collection = collection or VolumeCollection(appliance=appliance)
+        Navigatable.__init__(self, appliance=self.collection.appliance)
+
+    def wait_for_disappear(self, timeout=300):
+        def refresh():
+            self.provider.refresh_provider_relationships()
+            self.browser.refresh()
+
+        try:
+            wait_for(lambda: not self.exists,
+                     timeout=timeout,
+                     message='Wait for cloud Volume to disappear',
+                     delay=20,
+                     fail_func=refresh)
+        except TimedOutError:
+            logger.error('Timed out waiting for Volume to disappear, continuing')
+
+    def delete(self, wait=True):
+        """Delete the Volume"""
 
+        view = navigate_to(self, 'Details')
+        view.toolbar.configuration.item_select('Delete this Cloud Volume', handle_alert=True)
 
-@navigator.register(Volume, 'All')
+        view.entities.flash.assert_success_message('Delete initiated for 1 Cloud Volume.')
+
+        if wait:
+            self.wait_for_disappear(500)
+
+    @property
+    def exists(self):
+        view = navigate_to(self.collection, 'All')
+        try:
+            view.entities.get_entity(by_name=self.name)
```

Comment:
you need surf_pages=True here. Otherwise, it will fail if the volume isn't on current page

Reply:
Thanks, good point it will increase stability.